skimage/future/detect_obj/cascade.pyx

```diff
@@ -0,0 +1,832 @@
+# cython: boundscheck=False
+# cython: nonecheck=False
+# cython: wraparound=False
+# distutils: language = c++
+
+import numpy as np
+cimport numpy as cnp
+cimport openmp
+from libc.stdlib cimport malloc, free
+from libcpp.vector cimport vector
+from skimage._shared.transform cimport integrate
+
+# Use our own implementation of the function instead of libc.math.
+# Otherwise, the compilation breaks on some compilers due to Cython
+# bug.
+from skimage._shared.interpolation cimport round, fmax, fmin
+
+from cython.parallel import prange
+from skimage.color import rgb2gray
+from skimage.transform import integral_image
+import xml.etree.ElementTree as ET
+from ...feature._texture cimport _multiblock_lbp
+import math
+
+
+# Struct for storing clusters of rectangles. As the rectangles are dynamically
+# added, the sum of row, col positions and width and heights are stored
+# with the count of rectangles that belong to this cluster. This way,
+# we don't have to store all the rectangles information as array
+# and the average can be easily computed in a constant time.
+cdef struct DetectionsCluster:
+
+    int r_sum
+    int c_sum
+    int width_sum
+    int height_sum
+    int count
+
+
+# Struct for storing a single detection.
+cdef struct Detection:
+
+    int r
+    int c
+    int width
+    int height
+
+
+# Struct for storing multi-block binary pattern position.
+cdef struct MBLBP:
+
+    Py_ssize_t r
+    Py_ssize_t c
+    Py_ssize_t width
+    Py_ssize_t height
+
+
+# Struct for storing a stump of classifying cascade. It has the index to the
+# look-up table which is stored in Cascade class. Depending on the value of
+# the feature after its evaluation `left` or `right` value is returned which
+# is used by Cascade classifier to predict whether or not the
+# object is detected.
+cdef struct MBLBPStump:
+
+    Py_ssize_t feature_id
+    Py_ssize_t lut_idx
+    float left
+    float right
+
+
+# Struct for storing a stage of classifier which itself consists of stumps.
+# It has the index that maps to the starting stump and amount of stumps.
+# In each stage all the stumps are evaluated and their output values( `left`
+# or `right` depending on the input) are summed up and compared to the
+# threshold. If the value is higher than the threshold, the stage is passed
+# and Cascade classifier goes to the next one. If all the stages are passed,
+# the object is predicted to be present in the input image patch.
+cdef struct Stage:
+
+    Py_ssize_t first_idx
+    Py_ssize_t amount
+    float threshold
+
+
+cdef vector[Detection] _group_detections(vector[Detection] detections,
+                                         float intersection_score_threshold=0.5,
+                                         int min_neighbour_amount=4):
+    """Groups similar detection into single detection and eliminates weak
+    detections that have small amount of intersecting detection.
+
+    The function assumes that true detections are supposed to have a certain
```

Comment:
We assume that a true detection is characterized by a high number of overlapping detections.  Such detections are isolated and gathered into one cluster.  The average of each cluster is returned.

^^ Not sure what it means to "average a cluster"--so explain that as well.